kroxylicious-filters/kroxylicious-record-validation/src/main/java/io/kroxylicious/proxy/filter/schema/validation/bytebuf/JsonSchemaBytebufValidator.java

```diff
@@ -12,31 +12,90 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 
+import org.apache.kafka.common.header.internals.RecordHeaders;
 import org.apache.kafka.common.record.Record;
 
 import io.apicurio.registry.resolver.strategy.ArtifactReference;
+import io.apicurio.registry.serde.AbstractKafkaSerDe;
+import io.apicurio.registry.serde.DefaultIdHandler;
+import io.apicurio.registry.serde.IdHandler;
+import io.apicurio.registry.serde.headers.DefaultHeadersHandler;
+import io.apicurio.registry.serde.headers.HeadersHandler;
 import io.apicurio.schema.validation.json.JsonValidationResult;
 import io.apicurio.schema.validation.json.JsonValidator;
 
 import io.kroxylicious.proxy.filter.schema.validation.Result;
 
-public class JsonSchemaBytebufValidator implements BytebufValidator {
+import edu.umd.cs.findbugs.annotations.NonNull;
 
+public class JsonSchemaBytebufValidator implements BytebufValidator {
     private final JsonValidator jsonValidator;
+    private final Long globalId;
+    private final HeadersHandler keyHeaderHandler;
+    private final HeadersHandler valueHeaderHandler;
+    private final IdHandler keyIdHandler;
+    private final IdHandler valueIdHandler;
 
     public JsonSchemaBytebufValidator(Map<String, Object> schemaResolverConfig, Long globalId) {
+        this.globalId = globalId;
         this.jsonValidator = new JsonValidator(schemaResolverConfig, Optional.of(ArtifactReference.fromGlobalId(globalId)));
+        this.keyHeaderHandler = buildHeaderHandler(true);
+        this.keyIdHandler = buildIdHandler(true);
+
+        this.valueHeaderHandler = buildHeaderHandler(false);
+        this.valueIdHandler = buildIdHandler(false);
     }
 
     @Override
     public CompletionStage<Result> validate(ByteBuffer buffer, Record record, boolean isKey) {
         try {
-            JsonValidationResult jsonValidationResult = jsonValidator.validateByArtifactReference(buffer.clear());
+            // If the record includes a schema id, validate that it is consistent with the expected globalId.
```

Comment:
It would help if a commit and or PR description covered what we will now support. That is:

- globalId for key or value carried in a header using the default apicurio header names (`apicurio.key.globalId` or `apicurio.value.globalId`)
- globalId for key or value embedded in the message in apicurio's default 8-byte mode

And we should note it can be headers OR embedded ids.

The comment should probably be more explicit, using global id instead of schema id.